webapp/src/components/team_sidebar/team_sidebar.test.tsx

```diff
@@ -0,0 +1,49 @@
+import React from 'react';
+
+import {describe, expect, test, jest} from '@jest/globals';
+import {render} from '@testing-library/react';
+
+import TeamSidebar from './team_sidebar';
+
+const mockTheme = {
+    sidebarBg: '#ffffff',
+    sidebarText: '#333333',
+    sidebarUnreadText: '#ff0000',
+    sidebarTextHoverBg: '#eeeeee',
+    sidebarTextActiveBorder: '#007bff',
+    sidebarTextActiveColor: '#007bff',
+    sidebarHeaderBg: '#f8f9fa',
+    sidebarHeaderTextColor: '#495057',
+    onlineIndicator: '#28a745',
+    awayIndicator: '#ffc107',
+    dndIndicator: '#dc3545',
+    mentionBg: '#ffeb3b',
+    mentionBj: '#ffeb3b',
+    mentionColor: '#333333',
+    centerChannelBg: '#f8f9fa',
+    centerChannelColor: '#333333',
+    newMessageSeparator: '#007bff',
+    linkColor: '#007bff',
+    buttonBg: '#007bff',
+    buttonColor: '#ffffff',
+    errorTextColor: '#dc3545',
+    mentionHighlightBg: '#ffeb3b',
+    mentionHighlightLink: '#007bff',
+    codeTheme: 'solarized-dark',
+};
+
+jest.mock('../sidebar_buttons', () => ({
+    __esModule: true,
+    default: () => <div data-testid='mocked-sidebar-buttons'/>,
```

Comment:
I'm not sure I understand why this is necessary. What's the downside to having the component render the actual `SidebarButtons`?

Reply:
In my perspective, the responsibility of the `team_sidebar` component is to decide whether to render the buttons based on the show property. By mocking the `sidebar_buttons `component, I feel this introduces a separation of concerns that can prove  to be essential. 

Downside of the actual rendering of `sidebar_buttons`  component being,  we need to cover the all cases for the child component although we are testing the parent component.

I feel the actual rendering of `sidebar_buttons` can be done in a test of its own. wdyt ?

Reply:
@Raghunandhan8818 I agree with this. Though I personally think we should still render the tree as normal, to get some coverage on the child component and be able to notice differences in the default rendering of the child component, at our current state of "no component tests" in this particular project. IMO the benefits outweigh the drawbacks here. We don't really have a pattern of mocking _components_ in our projects, so this is just a bit unfamiliar to me. Although `team_sidebar` has only one job, it makes the snapshot more useful if we provide it more context in this case.

I'm mostly saying this because we don't have a test for `sidebar_buttons` at the moment, and I want to try to make up for that with this one snapshot test from the parent component. Any refactor of the `sidebar_buttons` component will show us rendering differences from its current state with this test.

Reply:
@hmhealey I'm curious what you think about this discussion on mocking child components in a component test

Reply:
If possible, I'd recommend avoiding that. One of the major benefits I've found with using Testing Library is that you can write the tests in an end-to-end style without having to actually run the E2E tests. It also makes it so that the tests don't break as soon as someone refactors the component and decides to move some code into a child component. I think those are why Testing Library always renders all the children of the component instead of giving you the option of shallowly rendering the component like Enzyme does.

That said, if there's some Context or something that we're missing to deeply render the component, I could understand starting with mocking for now. We've had to do a bunch of work to set up context in web app unit tests which has resulted in us adding [`renderWithContext`](https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/tests/react_testing_utils.tsx#L33) which covers most cases but still doesn't work for everything.

Relatedly, I'd also recommend not doing snapshot testing for something like this. I feel like most of the time when snapshot tests fail in the web app, people will just update them without looking to see why they failed. Even if we still want to mock the child component, I'd suggest doing it in a way that you can use `expect(...).toBeInTheDocument()` to confirm that the child is rendered instead of using a snapshot. That also means that people looking at the component in the future will understand more easily what to look for in the test results.

Sorry if this is all a bit of a knowledge dump. I feel like I need to write a testing manifesto to put in the developer docs 😅

Reply:
@Raghunandhan8818 So if there were more components rendered by `team_sidebar`, we would have to mock each one of those imports. To me, that's a sign that this testing approach can be brittle. I'm thinking we should render the tree as normal. What do you think?

Reply:
Sure @mickmister @hmhealey , lets do it that way. I have removed mocking from the team_sidebar.test and configured a mock store using @reduxjs/toolkit. I guess this is a good starting point. And further test cases can be added in team_sidebar.test.tsx to cover the sidebar_buttons entirely. But I feel, that can be taken as a seperate PR. wdyt ?

Reply:
> I feel like most of the time when snapshot tests fail in the web app, people will just update them without looking to see why they failed

Personally I highly value the snapshot tests, and take a deep look at them every time I review a PR that has them (as well as my own snapshot changes). The only downside imo is you have to update them when something about the DOM changes. Though the benefits of "seeing" that change in the PR is worth it imo.